text/0000-non-exhaustive.md

```diff
@@ -0,0 +1,390 @@
+marked as `#[doc(hidden)]`, and rustdoc should continue to emit this message.
+
+However, after this message (if any), it should offer an additional message
+saying "more variants may be added in the future," to clarify that the enum is
+non-exhaustive. It also hints to the user that in the future, they may want to
+fine-tune their match code to include future variants when they are added.
+
+These two messages should be distinct; the former says "this enum has stuff
+that you shouldn't see," while the latter says "this enum is incomplete and may
+be extended in the future."
+
+# How We Teach This
+
+Changes to rustdoc should make it easier for users to understand the concept of
+non-exhaustive enums in the wild.
+
+Additionally, in the chapter on enums, a section should be added specifically
+for non-exhaustive enums. Because error types are common in almost all crates,
+this case is important enough to be taught when a user learns Rust for the first
+time.
+
+# Drawbacks
+
+* The `#[doc(hidden)]` hack in practice is usually good enough.
+* An attribute may be more confusing than a dedicated syntax.
+* `non_exhaustive` may not be the clearest name.
+* It's unclear if this attribute should apply to other aspects of the language
+  than just enums.
+
+# Alternatives
+
+* Provide a dedicated syntax for enums instead of an attribute. This would
+  likely be done by adding a `..` variant, as proposed by the original
+  [extensible enums RFC][RFC 757].
+* Allow creating private enum variants, giving a less-hacky way to create a
+  hidden variant.
+* Document the `#[doc(hidden)]` hack and make it more well-known.
+
+# Unresolved questions
+
+Should there be a way to warn downstream crate users when their match is
+non-exuhaustive? What if a user wants to add a warning to their matches using
+non-exhaustive enums, to indicate that more code should be added to handle a new
+variant?
+
+Should the below extensions be added?
+
+## Extensions to structs
+
+It makes sense to eventually allow this attribute on structs as well. Unlike
+enums, it's very easy to make a struct non-exhaustive:
+
+```rust
+pub struct Thing {
+    pub number_of_eggs: usize,
+    pub friction_constant: f64,
+    pub can_swim: bool,
+    non_exhaustive: (),
+}
+```
+
+Because the `non_exhaustive` field is private, downstream crates cannot
+construct a value of `Thing` or exhaustively match its fields, even though they
+can access other fields. However, this seems less ergonomic than:
+
+```rust
+#[non_exhaustive]
+pub struct Thing {
+    pub number_of_eggs: usize,
+    pub friction_constant: f64,
+    pub can_swim: bool,
+}
+```
+
+Because then the upstream crate won't have to add the `non_exhaustive` field
+when constructing a value, or exhaustively matching patterns.
+
+## Extensions to traits
+
+Tangentially, it also makes sense to have non-exhaustive traits as well, even
+though they'd be non-exhaustive in a different way. Take this example from
+[`byteorder`]:
```

Comment:
You're right that they seem sufficiently different, but I see them all as solving the same problem: instead of preventing a user from implementing/using something by creating a hidden field, this simply states that not all of the dependencies can be listed by the user.